File: TestObjects.VB/VBCompiler/Binary.vb
Partial Module VBCompiler
    <TestObject(Binary)>
    Friend Add As Expression = IIFE(Function()
                                        Dim x As Double = 0, y As Double = 0
                                        Return Expr(Function() x + y)
                                    End Function)

    <TestObject(Binary)>
    Friend Divide As Expression = IIFE(Function()
                                           Dim x As Double = 0, y As Double = 0
                                           Return Expr(Function() x / y)
                                       End Function)

    <TestObject(Binary)>
    Friend Modulo As Expression = IIFE(Function()
                                           Dim x As Double = 0, y As Double = 0
                                           Return Expr(Function() x Mod y)
                                       End Function)

    <TestObject(Binary)>
    Friend Multiply As Expression = IIFE(Function()
                                             Dim x As Double = 0, y As Double = 0
                                             Return Expr(Function() x * y)
                                         End Function)

    <TestObject(Binary)>
    Friend Subtract As Expression = IIFE(Function()
                                             Dim x As Double = 0, y As Double = 0
                                             Return Expr(Function() x - y)
                                         End Function)

    <TestObject(Binary)>
    Friend AndBitwise As Expression = IIFE(Function()
                                               Dim i As Integer = 0, j As Integer = 0
                                               Return Expr(Function() i And j)
                                           End Function)

    <TestObject(Binary)>
    Friend OrBitwise As Expression = IIFE(Function()
                                              Dim i As Integer = 0, j As Integer = 0
                                              Return Expr(Function() i Or j)
                                          End Function)

    <TestObject(Binary)>
    Friend ExclusiveOrBitwise As Expression = IIFE(Function()
                                                       Dim i As Integer = 0, j As Integer = 0
                                                       Return Expr(Function() i Xor j)
                                                   End Function)

    <TestObject(Binary)>
    Friend AndLogical As Expression = IIFE(Function()
                                               Dim b1 As Boolean = True, b2 As Boolean = True
                                               Return Expr(Function() b1 And b2)
                                           End Function)

    <TestObject(Binary)>
    Friend OrLogical As Expression = IIFE(Function()
                                              Dim b1 As Boolean = True, b2 As Boolean = True
                                              Return Expr(Function() b1 Or b2)
                                          End Function)

    <TestObject(Binary)>
    Friend ExclusiveOrLogical As Expression = IIFE(Function()
                                                       Dim b1 As Boolean = True, b2 As Boolean = True
                                                       Return Expr(Function() b1 Xor b2)
                                                   End Function)

    <TestObject(Binary)>
    Friend [AndAlso] As Expression = IIFE(Function()
                                              Dim b1 As Boolean = True, b2 As Boolean = True
                                              Return Expr(Function() b1 AndAlso b2)
                                          End Function)

    <TestObject(Binary)>
    Friend [OrElse] As Expression = IIFE(Function()
                                             Dim b1 As Boolean = True, b2 As Boolean = True
                                             Return Expr(Function() b1 OrElse b2)
                                         End Function)

    <TestObject(Binary)>
    Friend Equal As Expression = IIFE(Function()
                                          Dim i As Integer = 0, j As Integer = 0
                                          Return Expr(Function() i = j)
                                      End Function)

    <TestObject(Binary)>
    Friend NotEqual As Expression = IIFE(Function()
                                             Dim i As Integer = 0, j As Integer = 0
                                             Return Expr(Function() i <> j)
                                         End Function)

    <TestObject(Binary)>
    Friend GreaterThanOrEqual As Expression = IIFE(Function()
                                                       Dim i As Integer = 0, j As Integer = 0
                                                       Return Expr(Function() i >= j)
                                                   End Function)

    <TestObject(Binary)>
    Friend GreaterThan As Expression = IIFE(Function()
                                                Dim i As Integer = 0, j As Integer = 0
                                                Return Expr(Function() i > j)
                                            End Function)

    <TestObject(Binary)>
    Friend LessThan As Expression = IIFE(Function()
                                             Dim i As Integer = 0, j As Integer = 0
                                             Return Expr(Function() i < j)
                                         End Function)

    <TestObject(Binary)>
    Friend LessThanOrEqual As Expression = IIFE(Function()
                                                    Dim i As Integer = 0, j As Integer = 0
                                                    Return Expr(Function() i <= j)
                                                End Function)

    <TestObject(Binary)>
    Friend Coalesce As Expression = IIFE(Function()
                                             Dim s1 As String = Nothing, s2 As String = Nothing
                                             Return Expr(Function() If(s1, s2))
                                         End Function)

    ' The VB compiler has special behavior around shift operations - https://docs.microsoft.com/dotnet/visual-basic/language-reference/operators/left-shift-operator#remarks
    <TestObject(Binary)>
    Friend LeftShift As Expression = IIFE(Function()
                                              Dim i As Integer = 0, j As Integer = 0
                                              Return Expr(Function() i << j)
                                          End Function)

    <TestObject(Binary)>
    Friend RightShift As Expression = IIFE(Function()
                                               Dim i As Integer = 0, j As Integer = 0
                                               Return Expr(Function() i >> j)
                                           End Function)

    <TestObject(Binary)>
    Friend ArrayIndex As Expression = IIFE(Function()
                                               Dim arr = New String() {17}
                                               Return Expr(Function() arr(0))
                                           End Function)

    <TestObject(Binary)>
    Friend Power As Expression = IIFE(Function()
                                          Dim x As Double = 0, y As Double = 0
                                          Return Expr(Function() x ^ y)
                                      End Function)
End Module

File: Tests/dynamicLinqTestObjects/Binary.cs
using ExpressionTreeTestObjects;
using System.Linq.Expressions;
using static ExpressionTreeTestObjects.Categories;

namespace ExpressionTreeToString.Tests {
    partial class DynamicLinqTestObjects {
        [TestObject(Binary)]
        internal static readonly Expression Add = Expr(p => p.Age + 100);

        [TestObject(Binary)]
        internal static readonly Expression Divide = Expr(p => p.Age / 2);

        [TestObject(Binary)]
        internal static readonly Expression Modulo = Expr(p => p.Age % 10);

        [TestObject(Binary)]
        internal static readonly Expression Multiply = Expr(p => p.Age * 10);

        [TestObject(Binary)]
        internal static readonly Expression Subtract = Expr(p => p.Age - 10);

        [TestObject(Binary)]
        internal static readonly Expression AndAlso = Expr(p => p.Age >= 20 && p.Age <= 60);

        [TestObject(Binary)]
        internal static readonly Expression OrElse = Expr(p => p.Age < 20 || p.Age > 60);

        [TestObject(Binary)]
        internal static readonly Expression Equal = Expr(p => p.LastName == "A");

        [TestObject(Binary)]
        internal static readonly Expression NotEqual = Expr(p => p.LastName != "A");

        [TestObject(Binary)]
        internal static readonly Expression GreaterThanOrEqual = Expr(p => p.Age >= 13);

        [TestObject(Binary)]
        internal static readonly Expression GreaterThan = Expr(p => p.Age > 65);

        [TestObject(Binary)]
        internal static readonly Expression LessThan = Expr(p => p.Age < 20);

        [TestObject(Binary)]
        internal static readonly Expression LessThanOrEqual = Expr(p => p.Age <= 20);

        [TestObject(Binary)]
        internal static readonly Expression Coalesce = Expr(p => p.LastName ?? "");

        [TestObject(Binary)]
        internal static readonly Expression ArrayIndex= Expr(p => p.Relatives[4]);

        [TestObject(Binary)]
        internal static readonly Expression LogicalAnd = Expr(p => p.Age >= 20 & p.Age <= 60);

        [TestObject(Binary)]
        internal static readonly Expression LogicalOr = Expr(p => p.Age >= 20 | p.Age <= 60);
    }
}
